sled-storage/src/manager.rs

```diff
@@ -933,17 +934,36 @@ impl StorageManager {
         log: &Logger,
         config: &DatasetsConfig,
     ) -> DatasetsManagementResult {
-        let mut status = vec![];
-        for dataset in config.datasets.values() {
-            status.push(self.dataset_ensure_internal(log, dataset).await);
-        }
+        // Gather properties about these datasets, if they exist.
+        //
+        // This pre-fetching lets us avoid individually querying them later.
+        let old_datasets = Zfs::get_dataset_properties(
+            config.datasets.values().map(|config| {
+                config.name.full_name()
+            }).collect::<Vec<String>>().as_slice(),
+            WhichDatasets::SelfOnly,
+        ).unwrap_or_default().into_iter().map(|props| {
+            (props.name.clone(), props)
+        }).collect::<BTreeMap<String, _>>();
+
+        // Ensure each dataset concurrently
+        let futures = config.datasets.values().map(|dataset| async {
+            self.dataset_ensure_internal(
+                log,
+                dataset,
+                old_datasets.get(&dataset.name.full_name()),
+            ).await
+        });
```

Comment:
I removed the `&mut self` constraint below -- it wasn't actually necessary -- and I'm ensuring all datasets concurrently now.